### ISSUE TITLE
Proposal: Move out DoneJS Dependencys to Generator

### ISSUE THREAD

**Issue by frank-dspeed** [donejs/cli#63]
I Engineered a lot on new versions of the donejs cli and donejs and that as i am a passionated developer my passion gave me the result that the depencys for add app shouldn't be in this project at all they should get added to the generator this way projects get easy update able and all that via installing a new generator.

this way we could move out all the code to generate project parts like add app plugin and that to the generator we give it a bin "add" this way we can strip out all parts from donejs cli that have to do with the generator and can then do donejs add and it will run the add binary from the generator-donejs.
this way we are more cool more stable and easyer update able 
we can also then implament other alternativ installers like Yarn more well 

greetings.

Hope that makes sense :D

**Comment by justinbmeyer** [comment]
@frank-dspeed Thanks for the passion!

Unfortunately, I'm not sure exactly what you mean.  I think something must not be coming through in the translation.  I cleaned up your comment to make more sense.

Having updated it, I'm still not very sure what you mean.  

We've already proposed moving some of the `donejs add app` behaviors that get installed into other `donejs add X` generators in this issue: https://github.com/donejs/donejs/issues/540

Are you recommending something other than that?

Some other thoughts that I'm not sure are relevant:

 - `yeoman` already exists, we don't need to be another generator framework.
 - DoneJS is an opinionated framework designed to provide a particular feature set.
 - you can build something like a DoneJS app manually via `npm install steal`, `npm install can`, etc


If you literally mean:

> dependencies for donejs add app shouldn't be in this project at all.

Dependencies should not be loaded from the generator.  DoneJS generators will take on a life independent from an application.  3 years from now, a donejs app might look very different.  Upgrading a generator will not be able to actually update the application.

**Comment by frank-dspeed** [comment]
@justinbmeyer 
At Present: all relativ from github.com:
donejs/donejs does handle donejs add via global bin 
-> it calls a internal init function that creates a dir and does installation of the donejs/cli
-> then it fires the current installed donejs/cli 
-> the donejs-cli will installIfMissing the generator and 
-> fire the generator.

MyProposal:
donejs/donejs installs a global yoman generator called generator-donejs
-> that gets used via yo donejs add
-> that can be then used also via donejs/donejs add translated to yo donejs add
-> the generator will create the project dir and install choosed donejs-cli in project and finish all the rest

so all configuration about the project goes into the generators

hope that makes sense

**Comment by frank-dspeed** [comment]
and by the way i have that ready i can upgrade the generator-donejs to work indipendent from donejs/donejs

**Comment by matthewp** [comment]
Hey @frank-dspeed a lot of this sounds really good.

It sounds like you are almost proposing getting rid of donejs-cli though, is that right?

One key thing that we did when designing the cli is make it so that if you install DoneJS 0.8, when you run `donejs add app` you will *always* get a 0.8 app.  We did this by making the donejs-cli versions match the major/minor versions of DoneJS.  Any proposed changes will need to account for that.

**Comment by frank-dspeed** [comment]
@matthewp they are accounting that at present and yes your right i am planing to put all functions into donejs global bin:

```
  New Design: donejs/donejs/bin/donejs
  can find current Project
  can run binarys localToProject
  can run generators localToProject
  can run global generators
  can installIfMissing localToProject
  can installIfMissingGlobal Global
  can run npm scripts localToProject
```

**Comment by frank-dspeed** [comment]
@justinbmeyer  @matthewp did i say it solves all existing issus listed here :) i only need your go then i stop working on my fork and rename all in it to donejs and make a pull request against github.com/donejs/donejs

and  github.com/donejs/generator-donejs

**Comment by frank-dspeed** [comment]
*** Update ***

```
# feathers-cli already uses yo and so we could also integrate directly into feathers project
warning feathers-cli > generator-feathers > cross-spawn-async@2.2.4: cross-spawn no longer requires a build toolchain, use it instead!
warning feathers-cli > yo > fullname > npmconf@2.1.2: this package has been reintegrated into npm and is now out of date with respect to npm
```

**Comment by phillipskevin** [comment]
I think that moving the responsibility for installing app dependencies into the app generator is a good idea. Currently, `donejs-cli` passes [this list](https://github.com/donejs/cli/blob/master/package.json#L62-L85) to the generators. This is used by [the app generator](https://github.com/donejs/generator-donejs/blob/d8f1a7f6ee39c8fe43ecba1c890399975273da5b/app/index.js#L165-L171) and the [plugin generator](https://github.com/donejs/generator-donejs/blob/36f09c1a2574c0693974e9e2d4d9699083599115/plugin/index.js#L165-L187) to write out a package.json file for the generated app/plugin.

I think the only reason for this is what @matthewp mentioned above (https://github.com/donejs/cli/issues/63#issuecomment-256079989).

But `generator-donejs` can behave the same way. If you have DoneJS 1.0 installed globally, running `donejs add app` will install `generator-donejs@1.0` so that the correct dependencies will be used. Or is there something I'm missing here? @matthewp @justinbmeyer

**Comment by frank-dspeed** [comment]
@phillipskevin yes your missing something :) but first of all :+1:  so now the Real Solution that i already did tested and i am using it in production since many month :).

**IMPORTANT NOTE GENERAL**: Yeoman generators are made to be installed Global! and designed to get updated via a internal auto update function of Yeoman

```
donejs-cli offers:
- using node_modules/.bin/* via donejs // this can easy be done via global donejs
- using generators via yeoman-encironment relativ to project/node_modules/ // can easy be done via global donejs but should be depricated because its agains Yeoman usage Best Practice
- It has a list of packages and versions right we can move that to the generator and so enable it to get installed global and it works like a charm also update able via Yeoman autoupgrade and i do version pining via a extra fild in package.json i store all my donejs settings in it in a { donejs: {} } Object
```

**Comment by matthewp** [comment]
@phillipskevin What would be the advantage of putting the list in the generator?

The idea of donejs-cli is that you shouldn't have to update the global donejs very often (just major version changes), and by having a local `node_modules/.bin/donejs` this will ensure that all members of a team are using the same donejs version regardless of what global version they are using.  The global donejs is meant to be a very thin layer that just calls into the local donejs.  Adding features to the global donejs would go against that.

Getting rid of donejs-cli in favor of generator-donejs doesn't solve that problem, donejs-cli is about more than just `donejs add app/plugin`.

**Comment by phillipskevin** [comment]
The purpose of moving the app dependencies to the app generator is just to keep the responsibility in one place. Also, I think generators should be responsible for installing their own dependencies.

This idea came about while working on `donejs-documentjs`. To remove `documentjs` from an app generated with `donejs`, I had to change `donejs-cli` to not install `documentjs` when creating a new app and change `generator-donejs` so that the `.md` and `documentjs.json` files are not created.

I thought about creating another issue for this since the discussion on this one is much broader than what I'm proposing, but I wasn't sure if that would be even more confusing since the title would be nearly identical.

**Comment by matthewp** [comment]
@phillipskevin Ok, if this is just about where to put the versioning information consider that the versions are by:

* `donejs add app`, which as you say would be more maintainable if it included the version in generator-donejs.
* Upgrades. Currently this is done manually by devs looking at the donejs-cli tag's package.json and seeing what to upgrade.  However we've talked about adding some type of `donejs update` or `donejs upgrade` utility, see https://github.com/donejs/donejs/issues/535

Given that the version numbers are used (potentially) by more than one tool, I think they should remain somewhere neutral. 

I wonder if it would make sense to have a `donejs-versions` repo which contains JSON files like `1.0.0.json`, etc. and then each tool would pull from that to get the version numbers of individual packages.

**Comment by phillipskevin** [comment]
Do developers really look at `donejs-cli` to see what to upgrade? I wouldn't have thought to do that.    When I was updating an app from 0.9 to 1.0, I created a fresh 1.0 app and compared the generated package.json to my existing 0.9 app package.json. Sort of similar to what Marshall mentioned here: https://github.com/donejs/donejs/issues/535#issuecomment-182963678.

The idea of a `donejs-versions` repo is interesting, although I think it would be difficult to test. If you update a value in `donejs-versions/1.0.0.json` and push a new release, you'd have to rely on CI running for all of the other tools that depend on this in order to see if the new version broke something.

**Comment by frank-dspeed** [comment]
i still think there should be a generator-donejs and additional versions in it and later splitted into moduls.

the generator has other ways to pin versions and detect installed versions even without a donejs-cli package.
